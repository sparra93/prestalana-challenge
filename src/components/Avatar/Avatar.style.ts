import styled from 'styled-components';

export const UserSection = styled.div`
  display: flex;
  align-items: center;
  padding-bottom: 20px;
  border-bottom: 1px solid #dddddd59;
`;

export const AvatarImg = styled.img`
  width: 50px;
  height: 50px;
  border-radius: 50%;
  margin-right: 10px;
`;

export const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  font-weight: 200;
`;

export const UserName = styled.span`
  font-weight: bold;
  font-size: 1em;
  font-weight: 700;
`;

export const UserEmail = styled.span`
  font-size: 0.9em;
`;

export const LogoutButton = styled.button`
  background-color: #ff4d4d;
  color: #fff;
  border: none;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: pointer;
  margin-top: 10px;

  &:hover {
    background-color: #ff1a1a;
  }
`;
