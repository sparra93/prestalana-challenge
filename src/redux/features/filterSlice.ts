import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import type { RootState } from '@/redux//store';

interface FilterState {
  yearFilter: string[];
  colorFilter: string[];
  pantoneFilter: string[];
  nameFilter: string;
}

const initialState: FilterState = {
  yearFilter: [],
  colorFilter: [],
  pantoneFilter: [],
  nameFilter: '',
};

const filterSlice = createSlice({
  name: 'filters',
  initialState,
  reducers: {
    setInitialYearFilter: (state, action: PayloadAction<string[]>) => {
      state.yearFilter = action.payload;
    },
    setInitialColorFilter: (state, action: PayloadAction<string[]>) => {
      state.colorFilter = action.payload;
    },
    setInitialPantoneFilter: (state, action: PayloadAction<string[]>) => {
      state.pantoneFilter = action.payload;
    },
    setYearFilter: (state, action: PayloadAction<string>) => {
      if (state.yearFilter.includes(action.payload)) {
        state.yearFilter = state.yearFilter.filter(
          (year) => year !== action.payload,
        );
      } else {
        state.yearFilter.push(action.payload);
      }
    },
    setColorFilter: (state, action: PayloadAction<string>) => {
      if (state.colorFilter.includes(action.payload)) {
        state.colorFilter = state.colorFilter.filter(
          (color) => color !== action.payload,
        );
      } else {
        state.colorFilter.push(action.payload);
      }
    },
    setPantoneFilter: (state, action: PayloadAction<string>) => {
      if (state.pantoneFilter.includes(action.payload)) {
        state.pantoneFilter = state.pantoneFilter.filter(
          (color) => color !== action.payload,
        );
      } else {
        state.pantoneFilter.push(action.payload);
      }
    },
    setNameFilter(state, action: PayloadAction<string>) {
      state.nameFilter = action.payload;
    },
  },
});

export const {
  setInitialYearFilter,
  setInitialColorFilter,
  setInitialPantoneFilter,
  setYearFilter,
  setNameFilter,
  setPantoneFilter,
  setColorFilter,
} = filterSlice.actions;
export const selectYearFilter = (state: RootState): string[] =>
  state.filter.yearFilter;
export const selectColorFilter = (state: RootState): string[] =>
  state.filter.colorFilter;
export const selectNameFilter = (state: RootState): string =>
  state.filter.nameFilter;
export const selectPantoneFilter = (state: RootState): string[] =>
  state.filter.pantoneFilter;

export default filterSlice.reducer;
