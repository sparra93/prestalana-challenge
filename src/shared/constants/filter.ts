import type { IFilterState } from '@/shared/Interfaces/Filter.interface';

export const filterSliceInitialState: IFilterState = {
  yearFilter: [],
  colorFilter: [],
  pantoneFilter: [],
  nameFilter: '',
};
