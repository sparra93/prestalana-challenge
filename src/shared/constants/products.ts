import type { IProductsState } from '../Interfaces/Product.interface';

export const productSliceInitialState: IProductsState = {
  products: [],
  cart: [],
  favorites: [],
  error: false,
  loading: false,
  loaded: false,
  cartCountTotal: 0,
};
