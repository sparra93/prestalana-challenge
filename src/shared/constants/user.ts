import type { IUserSliceInitialState } from '../Interfaces/User.interface';

export const userSliceInitialState: IUserSliceInitialState = {
  userInfo: null,
  error: false,
  loading: false,
  loaded: false,
  favorites: [],
};
