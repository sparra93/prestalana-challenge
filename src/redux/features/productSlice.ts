import type { PayloadAction } from '@reduxjs/toolkit';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import type { RootState } from '@/redux/store';
import { fetchProducts } from '@/services/products';
import { productSliceInitialState } from '@/shared/constants/products';
import {
  CART_NAME,
  EMAIL_NAME,
  FAVORITES_NAME,
} from '@/shared/constants/sessionStorage';
import {
  getFromLocalStorage,
  getFromSessionStorage,
  setLocalStorage,
} from '@/shared/helpers/storage';
import type {
  ICartProduct,
  IProduct,
  IProductsResponse,
} from '@/shared/Interfaces/Product.interface';

export const getAll = createAsyncThunk('products/getAll', async () => {
  const response = await fetchProducts();
  const data: IProductsResponse = await response.json();
  return data;
});

const productSlice = createSlice({
  name: 'products',
  initialState: productSliceInitialState,
  reducers: {
    addToCart(state, action: PayloadAction<IProduct>) {
      const { id } = action.payload;
      const existingItem = state.cart.find((item) => item.id === id);

      if (existingItem) {
        existingItem.quantity += 1;
      } else {
        state.cart.push({ ...action.payload, quantity: 1 });
      }

      state.cartCountTotal += 1;
      productSlice.caseReducers.setLocalStorageCart(state);
    },
    addToFavorites(state, action: PayloadAction<IProduct>) {
      const productToAdd = action.payload;
      const isDuplicate = state.favorites.some(
        (product) => product.id === productToAdd.id,
      );

      if (!isDuplicate) {
        state.favorites.push(productToAdd);
      }
      productSlice.caseReducers.setLocalStorageFavorites(state);
    },
    removeFromCart(state, action: PayloadAction<ICartProduct>) {
      state.cartCountTotal -= action.payload.quantity;
      state.cart = state.cart.filter(
        (product) => product.id !== action.payload.id,
      );
      productSlice.caseReducers.setLocalStorageCart(state);
    },
    decreaseQuantity(state, action: PayloadAction<ICartProduct>) {
      const newQuantity = action.payload.quantity - 1;
      const index = state.cart.findIndex(
        (product) => product.id === action.payload.id,
      );
      state.cart[index].quantity = newQuantity;
      if (newQuantity < 1) {
        state.cart = state.cart.filter(
          (product) => product.id !== action.payload.id,
        );
      }
      state.cartCountTotal -= 1;
      productSlice.caseReducers.setLocalStorageCart(state);
    },
    increaseQuantity(state, action: PayloadAction<ICartProduct>) {
      const newQuantity = action.payload.quantity + 1;
      const index = state.cart.findIndex(
        (product) => product.id === action.payload.id,
      );
      state.cart[index].quantity = newQuantity;
      state.cartCountTotal += 1;
      productSlice.caseReducers.setLocalStorageCart(state);
    },

    removeFromFavorites(state, action: PayloadAction<number>) {
      state.favorites = state.favorites.filter(
        (product) => product.id !== action.payload,
      );
      productSlice.caseReducers.setLocalStorageFavorites(state);
    },
    loadFavorites(state) {
      const favorites = getFromLocalStorage(FAVORITES_NAME) || '';
      const email = getFromSessionStorage(EMAIL_NAME) || '';
      if (email && favorites) {
        const userFavorites = !JSON.parse(favorites)[email]
          ? []
          : JSON.parse(favorites)[email];
        state.favorites = userFavorites;
      }
    },
    loadCart(state) {
      const cart = getFromLocalStorage(CART_NAME) || '';
      const email = getFromSessionStorage(EMAIL_NAME) || '';
      if (email && cart) {
        const userCart = !JSON.parse(cart)[email]
          ? []
          : JSON.parse(cart)[email];
        state.cart = userCart;
        state.cartCountTotal = state.cart.reduce(
          (accumulator, currentProduct) => {
            return accumulator + currentProduct.quantity;
          },
          0,
        );
      }
    },
    setLocalStorageFavorites: (state) => {
      const favorites = getFromLocalStorage(FAVORITES_NAME) || '';
      const email = getFromSessionStorage(EMAIL_NAME) || '';
      if (email) {
        const prev = favorites ? JSON.parse(favorites) : {};
        const newFavorites = {
          ...prev,
          [email]: state.favorites,
        };
        setLocalStorage(FAVORITES_NAME, JSON.stringify(newFavorites));
      }
    },
    setLocalStorageCart: (state) => {
      const cart = getFromLocalStorage(CART_NAME) || '';
      const email = getFromSessionStorage(EMAIL_NAME) || '';
      if (email) {
        const prev = cart ? JSON.parse(cart) : {};
        const newCart = {
          ...prev,
          [email]: state.cart,
        };
        setLocalStorage(CART_NAME, JSON.stringify(newCart));
      }
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getAll.pending, (state) => {
        state.loading = true;
        state.error = false;
      })
      .addCase(getAll.fulfilled, (state, action) => {
        state.loading = false;
        state.products = action.payload.data;
        state.favorites = [];
        state.loaded = true;
        productSlice.caseReducers.loadFavorites(state);
        productSlice.caseReducers.loadCart(state);
      })
      .addCase(getAll.rejected, (state) => {
        state.loading = false;
        state.error = true;
      });
  },
});

export const {
  addToCart,
  addToFavorites,
  removeFromCart,
  removeFromFavorites,
  decreaseQuantity,
  increaseQuantity,
} = productSlice.actions;
export const selectProducts = (state: RootState): IProduct[] =>
  state.products.products;
export const selectCart = (state: RootState): ICartProduct[] =>
  state.products.cart;
export const selectFavorites = (state: RootState): IProduct[] =>
  state.products.favorites;
export const selectCartCountTotal = (state: RootState): number =>
  state.products.cartCountTotal;

export default productSlice.reducer;
