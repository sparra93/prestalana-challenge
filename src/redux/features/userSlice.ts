import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import type { RootState } from '@/redux//store';
import { getUserList } from '@/services/auth';
import { EMAIL_NAME, TOKEN_NAME } from '@/shared/constants/sessionStorage';
import { userSliceInitialState } from '@/shared/constants/user';
import type { IProduct } from '@/shared/Interfaces/Product.interface';
import type {
  IUserInfo,
  IUserListResponse,
} from '@/shared/Interfaces/User.interface';

export const getUser = createAsyncThunk(
  'user/getUser',
  async (userEmail: string) => {
    const response = await getUserList();
    const userList: IUserListResponse = await response.json();
    const user = userList.data.find(({ email }) => email === userEmail);
    return user;
  },
);

const userSlice = createSlice({
  name: 'user',
  initialState: userSliceInitialState,
  reducers: {
    logout() {
      sessionStorage.removeItem(TOKEN_NAME);
      sessionStorage.removeItem(EMAIL_NAME);
      window.location.href = '/login';
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getUser.pending, (state) => {
        state.loading = true;
        state.error = false;
      })
      .addCase(getUser.fulfilled, (state, action) => {
        state.loading = false;
        state.userInfo = action.payload || null;
        state.loaded = true;
      })
      .addCase(getUser.rejected, (state) => {
        state.loading = false;
        state.error = true;
      });
  },
});

export const { logout } = userSlice.actions;
export const selectFavorites = (state: RootState): IProduct[] =>
  state.user.favorites;
export const selectCart = (state: RootState): IUserInfo | null =>
  state.user.userInfo;

export default userSlice.reducer;
