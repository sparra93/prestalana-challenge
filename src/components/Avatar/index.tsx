import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { logout } from '@/redux/features/userSlice';
import type { RootState } from '@/redux/store';

import {
  AvatarImg,
  LogoutButton,
  UserEmail,
  UserInfo,
  UserName,
  UserSection,
} from './Avatar.style';

interface AvatarProps {}

const Avatar: React.FC<AvatarProps> = () => {
  const userInfo = useSelector((state: RootState) => state.user.userInfo);
  const dispatch = useDispatch();
  const onClickLogout = () => {
    dispatch(logout());
  };
  return (
    <UserSection>
      <AvatarImg src={userInfo?.avatar} alt="User Avatar" />
      <UserInfo>
        <UserName>{userInfo?.first_name}</UserName>
        <UserEmail>{userInfo?.email}</UserEmail>
        <LogoutButton onClick={onClickLogout}>Cerrar Sesi√≥n</LogoutButton>
      </UserInfo>
    </UserSection>
  );
};

export default Avatar;
