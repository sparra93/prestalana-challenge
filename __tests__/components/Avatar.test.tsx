import { fireEvent, screen } from '@testing-library/react';
import React from 'react';

import Avatar from '@/components/Avatar';
import { logout } from '@/redux/features/userSlice';
import { userSliceInitialState } from '@/shared/constants/user';

describe('Avatar Component', () => {
  const customState = {
    user: {
      ...userSliceInitialState,
      userInfo: {
        id: 7,
        email: 'michael.lawson@reqres.in',
        first_name: 'Michael',
        last_name: 'Lawson',
        avatar: 'https://reqres.in/img/faces/7-image.jpg',
      },
    },
  };

  it('should display user information correctly', () => {
    renderWithProviders(<Avatar />, customState);

    expect(screen.getByAltText('User Avatar')).toHaveAttribute(
      'src',
      'https://reqres.in/img/faces/7-image.jpg',
    );
    expect(screen.getByText('Michael')).toBeInTheDocument();
    expect(screen.getByText('michael.lawson@reqres.in')).toBeInTheDocument();
  });

  it('should call logout when the logout button is clicked', () => {
    const { store } = renderWithProviders(<Avatar />);

    const logoutButton = screen.getByText('Cerrar SesiÃ³n');
    fireEvent.click(logoutButton);

    const actions = store.getActions();
    expect(actions).toContainEqual(logout());
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithProviders(<Avatar />, customState);

    expect(asFragment()).toMatchSnapshot();
  });
});
