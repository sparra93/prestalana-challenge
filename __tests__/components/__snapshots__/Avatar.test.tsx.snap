// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-2b2a1bd8-0 doDeWs"
  >
    <img
      alt="User Avatar"
      class="sc-2b2a1bd8-1 cVpnb"
      src="https://reqres.in/img/faces/7-image.jpg"
    />
    <div
      class="sc-2b2a1bd8-2 eVjdKf"
    >
      <span
        class="sc-2b2a1bd8-3 dbyjgu"
      >
        Michael
      </span>
      <span
        class="sc-2b2a1bd8-4 BLLUx"
      >
        michael.lawson@reqres.in
      </span>
      <button
        class="sc-2b2a1bd8-5 kfcqpw"
      >
        Cerrar Sesi√≥n
      </button>
    </div>
  </div>
</DocumentFragment>
`;
